File: core/src/extractor/token_splitter.rs
use shared::token::{token::{Token, TokenImpl}, token_type::TokenType};

pub fn split_tokens(
    tokens: &Vec<Token>,
    splitter: TokenType
) -> Vec<Vec<Token>> {
    let mut result: Vec<Vec<Token>> = Vec::new();
    let mut current: Vec<Token> = Vec::new();

    for token in tokens {
        if token.get_token_type() == splitter {
            if current.len() < 1 {
                continue;
            }

            result.push(current);
            current = Vec::new();
        } else {
            current.push(token.clone());
        }
    }

    if current.len() > 0 {
        result.push(current);
    }

    result
}

File: tokenUtil/tokenSplitter.go
package tokenUtil

import (
	"surf/logger"
	"surf/token"
)

// SplitTokens splits the given tokens
// using the given delimiter
func SplitTokens(
	tokens []token.Token,
	delimiter token.Type,
	nestedStartDelimiter token.Type,
	nestedEndDelimiter token.Type,
) [][]token.Token {
	result := make([][]token.Token, 0)
	current := make([]token.Token, 0)

	blockDepth := 0

	for _, unit := range tokens {
		tokenType := unit.GetType()

		if tokenType == nestedStartDelimiter {
			blockDepth++
		} else if tokenType == nestedEndDelimiter {
			blockDepth--

			if blockDepth < 0 {
				logger.TokenError(
					unit,
					"Unmatched delimiter",
					"Match the delimiters",
				)
			}
		}

		if tokenType == delimiter && blockDepth == 0 {
			if len(current) == 0 {
				logger.TokenError(
					unit,
					"Empty statement",
					"Add a statement to the list",
				)
			}

			result = append(result, current)
			current = make([]token.Token, 0)
			continue
		}

		current = append(current, unit)
	}

	if len(current) > 0 {
		result = append(result, current)
	}

	return result
}
